src/nvme/mi.h

```diff
@@ -3086,4 +3339,157 @@ int nvme_mi_admin_format_nvm(nvme_mi_ctrl_t ctrl,
+
+
+/**
+ * struct nvme_mi_aem_callbacks - Provided for nvme_mi_aem_enable
+ * @aem_handler: Callback function for application processing of events
+ * @enabled: Array indicating which AE should be enabled on the endpoint
+ *
+ * Application callbacks for nvme_mi_aem_process will be able to call
+ * nvme_mi_aem_get_next_event which will return a pointer to such an identifier
+ * for the next event the application should parse
+ */
+struct nvme_mi_aem_callbacks {
+	/*
+	 * This is called from inside nvme_mi_process when a payload has been validated and
+	 * can be parsed.  The application may call nvme_mi_aem_get_next_event from inside
+	 *  the callback to parse event data.
+	 */
+	enum nvme_mi_aem_handler_next_action (*aem_handler)(
+								nvme_mi_ep_t ep,
+								size_t num_events,
+								void *userdata);
+
+	bool enabled[256];
+};
+
+/**
+ * nvme_mi_aem_get_fd() - Returns the pollable fd for AEM data available
+ * @ep: The endpoint being monitored for asynchronous data
+ *
+ * This populated structure can be polled from the application to understand if
+ * a call to nvme_mi_aem_process() is required (when a poll returns > 0).
+ *
+ * Return: The fd value or -1 if error
+ */
+int nvme_mi_aem_get_fd(nvme_mi_ep_t ep);
+
+/**
+ * nvme_mi_aem_enable() - Enable AE on the provided endpoint
+ * @ep: Endpoint to enable AEs
+ * @envfa: Enable SR-IOV virtual functions AE
+ * @empfa: Enable SR-IOV physical functions AE
+ * @encfa: Enable PCIe functions AE
+ * @aemd: AEM Delay (time in seconds from when event happens to AEM being batched and sent)
+ * @aerd: AEM Retry Delay (time in 100s of ms between AEM retries from the endpoint)
+ * @callbacks: AE configuraiton including which events are enabled and the callback function
+ * @userdata: Application provided context pointer for callback function
+ *
+ * This function is called to enable AE on the endpoint.  Endpoint will provide initial state
+ * (if any) of enabled AEs and application can parse those via the aem_handler fn pointer in
+ * callbacks.  Thes can be obtained in the callback by calling nvme_mi_aem_get_next_event().
+ *
+ * Application should poll the fd that can be obtained from nvme_mi_aem_get_fd and then call
+ * nvme_mi_aem_process() when poll() indicates data available.
+ *
+ * A call to nvme_mi_aem_process() will grab AEM data and call the aem_handler fn pointer.
+ * At this point the application can call nvme_mi_aem_get_next_event() to get information for
+ * each triggered event.
+ *
+ * Return: 0 is a success, nonzero is an error and errno may be read for further details
+ */
+int nvme_mi_aem_enable(nvme_mi_ep_t ep,
+	bool envfa,
+	bool empfa,
+	bool encfa,
+	__u8 aemd,
+	__u8 aerd,
+	struct nvme_mi_aem_callbacks *callbacks,
+	void *userdata);
+
+
+/**
+ * nvme_mi_aem_get_enabled() - Return information on which AEs are enabled
+ * @ep: Endpoint to check enabled status
+ * @enabled: buffer indexed by AE event ID of enabled state
+ * @len: length of enabled.  Must be 256 bytes.
```

Comment:
The call semantics here are a bit odd; you're referring to `bytes` in the comment, but it's 256 booleans (which probably corresponds to 256 bytes, but that's not guaranteed).

We could define a more compact type for this perhaps; something like

```c
typedef uint64_t nvme_mi_aem_mask_t[4]
```

or if you prefer the boolean approach:

```c
typedef bool nvme_mi_aem_mask_t[256]
```

then drop the len argument, because you have a specific type instead.

Open to other suggestions though :)